Address post review comments for <https://commits.webkit.org/272703@main>

pvollan opened the pull request:
#### 90be7bc36004417e260f4f666c6411ad1591a6ff
<pre>
Address post review comments for &lt;<a href="https://commits.webkit.org/272703@main">https://commits.webkit.org/272703@main</a>&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=267275">https://bugs.webkit.org/show_bug.cgi?id=267275</a>
<a href="https://rdar.apple.com/120867428">rdar://120867428</a>

Reviewed by Jer Noble and Brent Fulgham.

This patch is addressing post review comments for &lt;<a href="https://commits.webkit.org/272703@main">https://commits.webkit.org/272703@main</a>&gt;. It also makes
sure that commit is called on the update coordinator in the UI process right after the message is sent
to the WebContent process to update the layer size.

Additionally, this patch addresses an issue where we did not update the hosting handle in the UI process
when the context ID changed. This issue would stop video rendering in certain cases.

* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
(WebCore::VideoPresentationModelVideoElement::setVideoSizeFenced):
* Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
(WebKit::RemoteMediaPlayerProxy::setVideoLayerSizeFenced):
* Source/WebKit/Platform/cocoa/LayerHostingContext.h:
(WebKit::LayerHostingContext::hostable const):
* Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
(WebKit::LayerHostingContext::createForExternalHostingProcess):
(WebKit::LayerHostingContext::~LayerHostingContext):
(WebKit::LayerHostingContext::contextID const):
(WebKit::LayerHostingContext::setFencePort):
(WebKit::LayerHostingContext::createHostingUpdateCoordinator):
(WebKit::LayerHostingContext::createHostingHandle):
(WebKit::LayerHostingContext::commit): Deleted.
* Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::createLayerHostViewWithID):
(WebKit::VideoPresentationManagerProxy::setVideoLayerFrame):
(-[WKLayerHostView dealloc]): Deleted.

Canonical link: <a href="https://commits.webkit.org/273143@main">https://commits.webkit.org/273143@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebe5c3d85dd9b1afa64ef072fdc54120b2723ed4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34439 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13256 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36439 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37125 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31146 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35526 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15646 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10367 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30145 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34958 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11217 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30709 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9804 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9889 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/30764 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38406 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31280 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31078 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/35959 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10010 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7891 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33903 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11821 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7911 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10563 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10864 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
```diff
@@ -215,13 +215,13 @@
     INFO_LOG_IF_POSSIBLE(LOGIDENTIFIER, size);
-    m_videoElement->setVideoLayerSizeFenced(size, WTFMove(fence));
+    m_videoElement->setVideoLayerSizeFenced(size, WTFMove(fence), WTFMove(completionHandler));
     m_videoElement->setVideoFullscreenFrame({ { }, size });
```
It seems weird to me that we would pass a completion handler to `setVideoLayerSizeFenced`, but then continue doing more work in this method (especially if `setVideoFullscreenFrame` has side-effects.

pvollan replied:
That is a good point. I have addressed this in the latest patch.

Thanks for reviewing!

brentfulgham commented on Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
```diff
@@ -99,20 +103,33 @@
     setVideoLayerSizeIfPossible(size);
 
-    m_player->setVideoLayerSizeFenced(size, WTFMove(machSendRight));
+    m_player->setVideoLayerSizeFenced(size, WTFMove(machSendRight), [] { });
```
Do we want to pass a no-op completion handler here? Does our completion handler not need to run after the player sets its layer size?

brentfulgham replied:
I wonder if the below code for committing the hosting update coordinator should be the completion handler for `m_player->setVideoLayerSizeFenced`.

pvollan replied:
You are completely right, this has been addressed in the latest version of the patch.

brentfulgham commented on Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
```diff
@@ -99,20 +103,33 @@
-    if (m_inlineLayerHostingContext)
-        m_inlineLayerHostingContext->commit();
+    [hostingUpdateCoordinator commit];
+    completionHandler();
```
We only use `completionHandler` if `USE(EXTENSIONKIT)` is true. I think this might create a compile error for platforms that don't use `EXTENSIONKIT`; you might need to add a `UNUSED_PARAM(completionHandler)` as an `#else` here.

brentfulgham replied:
Alternatively (and probably better style-wise) would be to call the completionHandler outside of the `#if USE(EXTENSIONKIT)` block.

brentfulgham replied:
Looking below at `MediaPlayerPrivateRemote::setVideoLayerSizeFenced`, it seems even more wrong that we pass a no-op completion handler, then run code that invokes the completion handler, since `setVideoLayerSizeFenced` clearly triggers async remote calls in some cases. I think that means we will sometimes call our completion handler before the async work completes, which doesn't seem right.

pvollan replied:
Yes, I believe you are right. I have addressed this and the compile issue in the latest version.

brentfulgham commented on Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
```diff
@@ -76,7 +76,8 @@
 #if USE(EXTENSIONKIT)
-    if (options.useHostable && [get_SEHostableClass() respondsToSelector:@selector(createHostableWithOptions:error:)]) {
+    if (options.useHostable) {
+        // Note: this reads like a leak, but the object returned from createHostableWithOptions should not be adopted.
```
This sounds like a very poorly named `create` function!

brentfulgham reviewed (commented):
I had a few questions about the patch, so I'm not comfortable approving quite yet (see below).

brentfulgham reviewed (approved):
Thank you for making those changes. I think this looks correct, though it would be best for an expert like Tim or Simon to review.

smfr commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -731,27 +722,26 @@ - (BOOL)prefersStatusBarHidden
     }
+
+#if USE(EXTENSIONKIT)
+    auto pid = m_page->process().processPool().gpuProcess()->processID();
```
It's not great to have to two ends of this communication (here and `RemoteMediaPlayerProxyCocoa`) be in two separate classes. It would be much nicer to factor all the `SEHosting`/xpc goop into its own class that both ends use.

pvollan replied:
That is a very good point! Will fix.

Thanks for reviewing!

brentfulgham reviewed (approved):
Looks good -- thanks for taking the time to address all of these comments! r+

jernoble reviewed (approved):
r=me, but the commit message should be updated to reflect the most recent implementation choices.

pvollan committed:
Address post review comments for <https://commits.webkit.org/272703@main>
https://bugs.webkit.org/show_bug.cgi?id=267275
rdar://120867428

Reviewed by Jer Noble and Brent Fulgham.

This patch is addressing post review comments for <https://commits.webkit.org/272703@main>. It also makes
sure that commit is called on the update coordinator in the UI process right after the message is sent
to the WebContent process to update the layer size.

Additionally, this patch addresses an issue where we did not update the hosting handle in the UI process
when the context ID changed. This issue would stop video rendering in certain cases.

* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
(WebCore::VideoPresentationModelVideoElement::setVideoSizeFenced):
* Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
(WebKit::RemoteMediaPlayerProxy::setVideoLayerSizeFenced):
* Source/WebKit/Platform/cocoa/LayerHostingContext.h:
(WebKit::LayerHostingContext::hostable const):
* Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
(WebKit::LayerHostingContext::createForExternalHostingProcess):
(WebKit::LayerHostingContext::~LayerHostingContext):
(WebKit::LayerHostingContext::contextID const):
(WebKit::LayerHostingContext::setFencePort):
(WebKit::LayerHostingContext::createHostingUpdateCoordinator):
(WebKit::LayerHostingContext::createHostingHandle):
(WebKit::LayerHostingContext::commit): Deleted.
* Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::createLayerHostViewWithID):
(WebKit::VideoPresentationManagerProxy::setVideoLayerFrame):
(-[WKLayerHostView dealloc]): Deleted.

Canonical link: https://commits.webkit.org/273143@main
